libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -33,6 +31,244 @@ _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
 #if defined(_CCCL_CUDA_COMPILER)
 
+template <class _Operand>
+using __cuda_atomic_enable_non_native_arithmetic =
+  __enable_if_t<_Operand::__size <= 16 || _Operand::__op == __atomic_cuda_operand::_f, bool>;
+
+template <class _Operand>
+using __cuda_atomic_enable_non_native_bitwise = __enable_if_t<_Operand::__size <= 16, bool>;
+
+template <class _Operand>
+using __cuda_atomic_enable_native_bitwise = __enable_if_t<_Operand::__size >= 32, bool>;
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void
+__cuda_atomic_load(const _Type* __ptr, _Type& __dst, _Order, _Operand, _Sco, __atomic_cuda_mmio_disable)
+{
+  constexpr uint64_t __alignmask = (sizeof(uint32_t) - 1);
+  uint32_t* __aligned            = (uint32_t*) ((intptr_t) __ptr & (~__alignmask));
+  const uint8_t __offset         = uint32_t((intptr_t) __ptr & __alignmask) * 8;
+
+  uint32_t __value = 0;
+
+  __cuda_atomic_load(__aligned, __value, _Order{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});
+  __dst = static_cast<_Type>(__value >> __offset);
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE bool
+__cuda_atomic_compare_exchange(_Type* __ptr, _Type& __dst, _Type __cmp, _Type __op, _Order, _Operand, _Sco)
+{
+  constexpr uint64_t __alignmask = (sizeof(uint32_t) - 1);
+  constexpr uint32_t __sizemask  = (1 << (sizeof(_Type) * 8)) - 1;
+  uint32_t* __aligned            = (uint32_t*) ((intptr_t) __ptr & (~__alignmask));
+  const uint8_t __offset         = uint32_t((intptr_t) __ptr & __alignmask) * 8;
+  const uint32_t __valueMask     = __sizemask << __offset;
+  const uint32_t __windowMask    = ~__valueMask;
+  const uint32_t __cmpOffset     = __cmp << __offset;
+  const uint32_t __opOffset      = __op << __offset;
+
+  // Algorithm for 8b CAS with 32b intrinsics
```

Comment:
On CTKs with `atom.cas.b16` available, this should probably use that instead of `atom.cas.b32`.

Reply:
I'd like to save that for later. It won't be impossible, but will require some rather difficult embedding of `#ifdef` around what is available and what is not.